src/main.rs

```diff
@@ -1,3 +1,88 @@
+use git2::{Commit, Error, Time};
+use git_graph::graph::GitGraph;
+
+struct Args {}
+
 fn main() {
-    println!("Hello, world!");
+    let args = Args {};
+    match run(&args) {
+        Ok(()) => {}
+        Err(e) => println!("error: {}", e),
+    }
+}
+
+fn run(_args: &Args) -> Result<(), Error> {
+    let path = ".";
+    let graph = GitGraph::new(path)?;
+    for info in &graph.commits {
+        let commit = &graph.commit(info.oid)?;
+        print_commit_short(commit, &info.branches);
+    }
+    Ok(())
+}
+
+#[allow(dead_code)]
+fn print_commit_short(commit: &Commit, branches: &Vec<String>) {
+    let symbol = if commit.parents().len() > 1 {
+        "\u{25CB}"
+    } else {
+        "\u{25CF}"
+    };
+    let branch_str = if branches.is_empty() {
+        "".to_string()
+    } else {
+        format!(" ({})", branches.join(", "))
+    };
+
+    let str = format!(
+        "{} {}{} {}",
+        symbol,
+        &commit.id().to_string()[0..7],
+        branch_str,
+        &commit.summary().unwrap_or("---")
+    );
+    println!("{}", str);
```

Comment:
Why the intermediate `str` variable instead of `println!` directly (as both will go through exactly the same formatting machinery).

Reply:
Good point!